app/src/main/kotlin/org/cru/godtools/ui/dashboard/lessons/LessonsViewModel.kt

```diff
@@ -1,25 +1,115 @@
 package org.cru.godtools.ui.dashboard.lessons
 
+import android.content.Context
+import androidx.lifecycle.SavedStateHandle
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import dagger.hilt.android.lifecycle.HiltViewModel
+import dagger.hilt.android.qualifiers.ApplicationContext
 import javax.inject.Inject
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.flatMapLatest
 import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.flow.stateIn
 import org.cru.godtools.analytics.model.OpenAnalyticsActionEvent
 import org.cru.godtools.analytics.model.OpenAnalyticsActionEvent.Companion.ACTION_OPEN_LESSON
 import org.cru.godtools.analytics.model.OpenAnalyticsActionEvent.Companion.SOURCE_LESSONS
+import org.cru.godtools.base.Settings
+import org.cru.godtools.db.repository.LanguagesRepository
 import org.cru.godtools.db.repository.ToolsRepository
+import org.cru.godtools.db.repository.TranslationsRepository
+import org.cru.godtools.model.Language
+import org.cru.godtools.model.Language.Companion.filterByDisplayAndNativeName
+import org.cru.godtools.ui.dashboard.tools.ToolsScreen.Filters.Filter
 import org.greenrobot.eventbus.EventBus
 
+const val KEY_SAVED_LESSON_LANGUAGE_LOCALE = "savedLessonLanguageLocale"
+
 @HiltViewModel
-class LessonsViewModel @Inject constructor(private val eventBus: EventBus, toolsRepository: ToolsRepository) :
-    ViewModel() {
-    val lessons = toolsRepository.getLessonsFlow()
-        .map { it.filterNot { it.isHidden }.sortedBy { it.defaultOrder }.mapNotNull { it.code } }
+class LessonsViewModel @Inject constructor(
+    @ApplicationContext
+    private val context: Context,
+    private val eventBus: EventBus,
+    toolsRepository: ToolsRepository,
+    languagesRepository: LanguagesRepository,
+    private val translationsRepository: TranslationsRepository,
+    private val savedStateHandle: SavedStateHandle,
+    val settings: Settings,
+) : ViewModel() {
+    val selectedLanguage = languagesRepository.getLanguagesFlow().combine(
+        savedStateHandle.getStateFlow(
+            KEY_SAVED_LESSON_LANGUAGE_LOCALE,
+            settings.appLanguage
+        )
+    ) { languages, locale ->
+        languages.findLast { it.code == locale } ?: Language(settings.appLanguage)
+    }
```

Comment:
you could have also done:

```kotlin
val selectedLanguage = savedStateHandle.getStateFlow(KEY_SAVED_LESSON_LANGUAGE_LOCALE, settings.appLanguage)
    .flatMapLatest { locale ->
        languagesRepository.findLanguageFlow(locale)
            .map { it ?: Language(locale) }
    }
```

Reply:
this still need to be changed

Reply:
instead of loading all languages, then searching for the language you want, you should just directly fetch the language you want